### PR TITLE
KubeVirt API endpoints for instancetypes and preferences

### LINKED ISSUE #11032: API endpoints for instancetype and preference

### PR DESCRIPTION
**Pull request by hdurand0710**
Signed-off-by: Helene Durand <helene@kubermatic.com>

**What this PR does / why we need it**:
This PR adds new endpoints for KubeVirt provider:
- list instancetypes, which returns the list of VirtualMachineInstancetype (kubermatic _standard_) concatenated with VirtualMachineClusterInstancetype (_custom_)
- list preferences, which returns the list of VirtualMachinePreference (kubermatic _standard_) concatenated with VirtualMachineClusterPreference (_custom_)

**Which issue(s) this PR fixes**:

Fixes #11032 

**What type of PR is this?**

/kind feature
/kind api-change

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
KubeVirt: add new API endpoints to list instancetypes and preferences.
```

**Documentation**:

```documentation
TBD. Documentation will added when the last PR for the Epic is raised, once the exact format of the `RawConfig` is decided. 
```

### CODE REVIEW

File: pkg/handler/v2/provider/kubevirt.go [hunk 1]
```diff
@@ -42,31 +42,61 @@ type KubeVirtGenericReq struct {
...
 
 // KubeVirtGenericNoCredentialReq represent a generic KubeVirt request with cluster credentials.
-// swagger:parameters listKubeVirtVMIPresetsNoCredentials listKubevirtStorageClassesNoCredentials
+// swagger:parameters listKubeVirtVMIPresetsNoCredentials listKubevirtStorageClassesNoCredentials listKubeVirtInstancetypesNoCredentials listKubeVirtPreferencesNoCredentials
 type KubeVirtGenericNoCredentialReq struct {
 	cluster.GetClusterReq
 }
 
+func kubeconfigFromRequest(ctx context.Context, request interface{}, presetsProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter) (string, error) {
+	req := request.(KubeVirtGenericReq)
```
Comment (pkprzekwas) [inline pkg/handler/v2/provider/kubevirt.go]: I suggest adding here a check for casting type.
```
req, ok := request.(KubeVirtGenericReq)
```

Comment (hdurand0710) [reply pkg/handler/v2/provider/kubevirt.go]: done

File: pkg/handler/common/provider/kubevirt_test.go [hunk 2]
```diff
@@ -0,0 +1,193 @@
+/*
+Copyright 2021 The Kubermatic Kubernetes Platform contributors.
```
Comment (mfranczy) [inline pkg/handler/common/provider/kubevirt_test.go]: 2022 :)

File: pkg/handler/v2/provider/kubevirt.go [hunk 3]
```diff
@@ -42,60 +42,106 @@ type KubeVirtGenericReq struct {
...
+		}
+		if credentials := preset.Spec.Kubevirt; credentials != nil {
+			kubeconfig = credentials.Kubeconfig
+		}
+	}
+	return kubeconfig, nil
+}
+
 // KubeVirtVMIPresetsEndpoint handles the request to list available KubeVirtVMIPresets (provided credentials).
 func KubeVirtVMIPresetsEndpoint(presetsProvider provider.PresetProvider, userInfoGetter provider.UserInfoGetter, settingsProvider provider.SettingsProvider) endpoint.Endpoint {
```
Comment (mfranczy) [inline pkg/handler/v2/provider/kubevirt.go:76]: please mark `KubeVirtVMIPresetsEndpoint` as deprecated

File: pkg/handler/v2/provider/kubevirt.go [hunk 4]
```diff
@@ -42,60 +42,106 @@ type KubeVirtGenericReq struct {
...
 	return func(ctx context.Context, request interface{}) (interface{}, error) {
-		req := request.(KubeVirtGenericReq)
-		kubeconfig := req.Kubeconfig
+		kubeconfig, err := kubeconfigFromRequest(ctx, request, presetsProvider, userInfoGetter)
```
Comment (mfranczy) [inline pkg/handler/v2/provider/kubevirt.go:78]: @lsviben what's the difference between requests with `WithClusterCredentialsEndpoint` sufix and without?? I see in both cases we have access to credentials. I am trying to understand the reason why we have those two different endpoints.

File: pkg/handler/v2/provider/kubevirt.go [hunk 5]
```diff
@@ -42,60 +42,106 @@ type KubeVirtGenericReq struct {
...
+		if err != nil {
+			return nil, err
 		}
-		return providercommon.KubeVirtVMIPresets(ctx, kubeconfig, nil, settingsProvider)
+		return providercommon.KubeVirtPreferences(ctx, kubeconfig, nil, settingsProvider)
 	}
 }
 
 // KubeVirtVMIPresetsWithClusterCredentialsEndpoint handles the request to list available KubeVirtVMIPresets (cluster credentials).
 func KubeVirtVMIPresetsWithClusterCredentialsEndpoint(projectProvider provider.ProjectProvider, privilegedProjectProvider provider.PrivilegedProjectProvider, seedsGetter provider.SeedsGetter, userInfoGetter provider.UserInfoGetter, settingsProvider provider.SettingsProvider) endpoint.Endpoint {
```
Comment (mfranczy) [inline pkg/handler/v2/provider/kubevirt.go:111]: please mark as deprecated

File: pkg/handler/v2/provider/kubevirt.go (hunk 4 above)
Comment (hdurand0710) [reply pkg/handler/v2/provider/kubevirt.go:78]: From my understanding, @lsviben to confirm.
1) Without the `WithClusterCredentialsEndpoint` is used in the cluster creation flow, before the cluster and credentials secrets are created.
Request header contains the credentials: the KubeVirt kubeconfig or the Kubermatic preset containing the kubeconfig.
2) With the WithClusterCredentialsEndpoint is used once the Cluster is created and credentials stored in the appropriate secret in the `kubermatic` namespace.
`/api/v2/projects/{project_id}/clusters/{cluster_id}/providers/kubevirt/instancetypes`, credentials are not passed in the request, but retrieved from the secret.

File: pkg/handler/v2/provider/kubevirt.go (hunk 5 above)
Comment (hdurand0710) [reply pkg/handler/v2/provider/kubevirt.go:111]: done

File: pkg/handler/v2/provider/kubevirt.go (hunk 3 above)
Comment (hdurand0710) [reply pkg/handler/v2/provider/kubevirt.go:76]: done

File: pkg/handler/v2/provider/kubevirt.go (hunk 4 above)
Comment (lsviben) [reply pkg/handler/v2/provider/kubevirt.go:78]: @hdurand0710 is correct :)

Comment (mfranczy) [reply pkg/handler/v2/provider/kubevirt.go:78]: thanks for detailed explanation

File: pkg/handler/v2/provider/kubevirt.go (hunk 5 above)
Comment (mfranczy) [reply pkg/handler/v2/provider/kubevirt.go:111]: could you fine-tune this a little bit that it's being parsed by all IDEs correctly? Please follow the convention described here https://github.com/golang/go/wiki/Deprecated, meaning it should look like this:

```
// KubeVirtVMIPresetsWithClusterCredentialsEndpoint handles the request to list available KubeVirtVMIPresets (cluster credentials).
//
// Deprecated: in favor of KubeVirtInstancetypesWithClusterCredentialsEndpoint.
```

Comment (hdurand0710) [reply pkg/handler/v2/provider/kubevirt.go:111]: sure, done. 
Note that I love the kindness of your wording: "fine-tune this a little bit" :-)  how could I refuse ?

Comment (mfranczy) [reply pkg/handler/v2/provider/kubevirt.go:111]: :-)

**Review by mfranczy — APPROVED** [review]
/approve
/lgtm

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve 🎊

**Pull request merged**